client/layout/guided-tours/docs/ARCHITECTURE-FUTURE.md

```diff
@@ -0,0 +1,150 @@
+
+## Making `actionLog` scale
+
+By design, `actionLog`'s reducer will react to all sorts of actions coming in through the dispatcher. As of this writing, the [list][relevant-types] of subscribed types is relatively limited:
+
+```
+COMPONENT_INTERACTION_TRACKED
+FIRST_VIEW_HIDE
+GUIDED_TOUR_UPDATE
+THEMES_RECEIVE
+PREVIEW_IS_SHOWING
+ROUTE_SET
+SITE_SETTINGS_RECEIVE
+```
+
+Yet, a few things are clear:
+
+- As more Guided Tours are built in the future, that list can only grow.
+- There are really no limits to _what kind_ of actions may be added to it — from navigation, to success signals of fetches, to specific signals of user interaction.
+- A non-negligible part of these actions can be fired very often (`ROUTE_SET`, `COMPONENT_INTERACTION_TRACKED`).
+
+Every time `actionLog` changes, Guided Tours's main selectors get called with the new state, whether a tour is running or not. This has obvious **performance implications**. Right now, we believe this is acceptably mitigated by 1) copious use of memoization with `createSelector` and simple cache heuristics, and 2) the still limited set of action types `actionLog` subscribes to.
+
+Nevertheless, whenever the selectors have to be fully reevaluated, all of `actionLog` has to be considered anew: all those sub-selectors will perform multiple `map`s, `reduce`s, etc. on the latest `actionLog`, even though most of its items have already been mapped, reduced, etc. in previous versions of `actionLog`.
+
+Now, you know what they say about premature optimization. Which is why we haven't optimized yet. But this document seeks to gather thoughts for the day when, through meticulous profiling (best-case scenario) or out of desperate hunches (worst-), we decide optimization is needed. Next comes a collection of ideas on how to address the issue, some of them composable, some mutually incompatible.
+
+### Streams!
+
+You probably saw this one coming. Calypso does not use [streams][rxjs], but, should we decide in favor of them, an improved `actionLog` would be had at minimum cost: as mentioned in _§ On the nature of `actionLog`_, the RxJS operators would naturally be adequate and optimized to handle `actionLog`. Realistically, though, adopting RxJS is much more than just pulling a large library (compressed/uncompressed: 30.1/138 KB), it's adopting a whole new programming paradigm. As much as I'd like to see that happen, it probably won't anytime soon.
+
+### (Persistent?) Data Structures
+
+This approach is a poor man's substitute for streams. It aims to replace standard array traversal functions with a suite of functions optimized for `actionLog`: `mapActionLog`, `filterActionLog`, etc. Since `actionLog` is a [monoid][monoids], any operation _f_ on `actionLog` can be decomposable to [divide and conquer]. Notably,
```

Comment:
Could you rephrase? "Using persistent data structures could be a simple replacement for streams ..."